controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,418 @@
+	"github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/apis/status/v1beta1"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/utils/strings/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	ControllerName       = "constraintstatus_reconciler"
+	ErrNotFoundDiscovery = errors.New("There is no matched apiGroup, version or kind")
+)
+
+type discoveryInfo struct {
+	apiResourceList        []*metav1.APIResourceList
+	discoveryLastRefreshed time.Time
+}
+
+type ConstraintStatusReconciler struct {
+	client.Client
+	Log           logr.Logger
+	DynamicClient dynamic.Interface
+	ClientSet     *kubernetes.Clientset
+	Namespace     string
+	constraintMap sync.Map
+	discoveryInfo
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ConstraintStatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// The work queue prevents the same item being reconciled concurrently:
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/1416#issuecomment-899833144
+		WithOptions(controller.Options{MaxConcurrentReconciles: int(5)}).
+		// watch Config resrouce as secondary
+		Named(ControllerName).
+		For(&v1beta1.ConstraintPodStatus{},
+			builder.WithPredicates(predicate.Funcs{
+				// Execute this reconcile func when it is audit-constraintStatuspod
+				// because a constraint creates 4 constraintStatuspods
+				CreateFunc: func(e event.CreateEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldObj := e.ObjectOld.(*v1beta1.ConstraintPodStatus)
+					newObj := e.ObjectNew.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(newObj.Status.Operations, "audit") &&
+						oldObj.Status.ObservedGeneration != newObj.Status.ObservedGeneration
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect requested constraints and add found kinds which is used in constraint to config.spec.sync.syncOnly
+func (r *ConstraintStatusReconciler) Reconcile(ctx context.Context,
```

Comment:
Sorry, I didn't explain well in my last PR review the algorithm I had in mind. Please consider the following algorithm:
* `constraintMap` (should be renamed to `constraintToSyncOnly` or similar) is a map where the keys are `ConstraintPodStatus` names (I don't think they are namespaced so namespace isn't required). The values are the syncOnly entries the associated constraint requires.
* Have `constraintMap` be prepopulated before the reconciler starts or ensure the first reconcile is serial to populate `constraintMap`. One idea is to initialize a channel as a struct member for `ConstraintStatusReconciler` with a size of 1 and include a single value such as an empty struct. The `Reconcile` method starts with a select statement with the case to get the value from the channel and the case to check if the channel is closed. The one goroutine that gets the empty struct will populate `constraintMap` entirely, close the channel, and return. Then the select statement for the closed channel case would just break out of the select statement and perform the normal reconcile so that way all other reconciles will assume the map is prepopulated.
* A reconcile request occurs. If the `ConstraintPodStatus` is deleted, remove the key for that `ConstraintPodStatus` in `constraintMap` and recompute the syncOnly based on all the values. If the value is different than what is on the `Config` object, then update it.
* A reconcile request occurs. If the `ConstraintPodStatus` is added or updated, set the syncOnly entries to `constraintMap` and recompute the syncOnly based on all the values. If the value is different than what is on the `Config` object, then update it.

Reply:
@yiraeChristineKim another thought, lets disable concurrency on this controller. I don't think you'll get much benefit of having it concurrent and it makes the code more complex.

Reply:
I expected `constraintMap` type is` sync.Map` what is Channel. I understand 3 bullets except second one